### PR TITLE
feat: remove custom institutions, enforce PDP / Edvise / Legacy in API and uploads

### PR DESCRIPTION
**Pull request by chapmanhk**
## changes

- **Institutions API (`routers/institutions.py`)**
  - Remove the custom-institution model from create/update semantics: every institution must resolve to **exactly one** of PDP (`pdp_id`), Edvise Schema / ES (`edvise_id` or `is_edvise` with auto-id), or Legacy (`legacy_id` or `is_legacy` with auto-id).
  - **POST:** Validates mutual exclusivity and “exactly one type”; idempotent duplicate POST (same name + state) **returns 400** if the existing row is misconfigured (e.g. more than one id set, or none).
  - **PATCH:** Same invariants; `is_edvise` / `is_legacy` can auto-assign ids like POST; **`schemas` is recomputed only when the school-type triple changes**; `allowed_schemas` alone can still replace `schemas` when the type does not change.
  - **Refactor:** Shared error copy (`_MUTUALLY_EXCLUSIVE_SCHOOL_TYPE_DETAIL`), helpers for duplicate-post checks, PATCH merge/validate/persist, and a slimmer `update_inst` (DRY + structure aligned with internal style guidelines).

- **Data / upload path (`routers/data.py` and related)**
  - Upload resolution limited to **edvise / pdp / legacy** (no custom-institution path).

- **Utilities**
  - Legacy schema grouping updated (e.g. `LEGACY_SCHEMA_GROUP` includes `UNKNOWN` as appropriate for this model).

- **Tests (`institutions_test.py`, and related where touched)**
  - Coverage for duplicate POST edge cases, PATCH flags, `allowed_schemas`-only PATCH, and invalid existing rows.

## context

Custom institutions are being retired. The API and upload pipeline must not accept or branch on a fourth “custom” school type; institutions and uploads must map cleanly to **PDP, Edvise (ES), or Legacy** with consistent validation and predictable `schemas` behavior. Hardening duplicate POST avoids silently succeeding when a legacy bad row exists (typeless or conflicting ids).

https://app.asana.com/1/6325821815997/project/1208486153653829/task/1213791624909790?focus=true

## testing
Tested deployed feature branch on dev with its associated edvise-ui feature branch: confirmed that custom is removed and that school creation is working as intended.

## questions

- None

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1213998619119545

### CODE REVIEW

File: src/webapp/utilities.py [hunk 1]
```diff
@@ -153,8 +153,7 @@ class SchemaType(StrEnum):
 }
 
 LEGACY_SCHEMA_GROUP: Final = {
-    SchemaType.STUDENT,
-    SchemaType.COURSE,
+    SchemaType.UNKNOWN,
```
Comment (vishpillai123) [inline src/webapp/utilities.py:156]: Do we want to use "UNKNOWN" as a catch-all for any type of dataset?

Is it possible to remove LEGACY_SCHEMA_GROUP entirely?

File: src/webapp/validation_extension.py
Comment (vishpillai123) [inline src/webapp/validation_extension.py:1]: Given that we are removing schema extensions, which is the right call to me given how we're evolving in Edvise, I'm thinking we'll have to reconsider or even deprecate our previous API validation layer (Mesh's validation implementation, not what you've recently implemented for PDP). This is out of scope for this PR, but curious to hear your thoughts here.

File: src/webapp/routers/institutions.py [hunk 2]
```diff
@@ -39,6 +39,17 @@
     tags=["institutions"],
 )
 
+# PATCH/POST: every institution must resolve to exactly one school type (IDs on the row).
+_EXACTLY_ONE_SCHOOL_TYPE_DETAIL = (
+    "Institution must be exactly one of PDP (set pdp_id), "
+    "Edvise Schema (ES) (set edvise_id or is_edvise), or Legacy "
+    "(set legacy_id or is_legacy)."
+)
+_MUTUALLY_EXCLUSIVE_SCHOOL_TYPE_DETAIL = (
```
Comment (vishpillai123) [inline src/webapp/routers/institutions.py:48]: thank you for building this part out!

**Comment by vishpillai123** [comment]
what is a school-type triple?

File: src/webapp/utilities.py (hunk 1 above)
Comment (chapmanhk) [reply src/webapp/utilities.py:156]: When a file is validated, the code must always get at least one schema name in the list it validates against. An empty list causes failure. For legacy schools, we still save a default UNKNOWN on the institution that way the stored schemas field is never empty.
So we need something for LEGACY_SCHEMA_GROUP schemas, I just made it "UNKNOWN". If we want to call it something different I can adjust it.

File: src/webapp/validation_extension.py
Comment (chapmanhk) [reply src/webapp/validation_extension.py:1]: So kind of - technically legacy validation is still set up using the schema validation Mesh set up, we're just basically using what's already been built to say that if something is a legacy school it doesn't need any specific validations applied to it.
So once we have Edvise also on the Pandera validation system, legacy will be the only thing using the old system. At that point it would probably make sense to reconfigure/refactor validation for legacy schools so that we can simplify the code (since we're not actually using it for what it was supposed to be used for). So agreed on this - but technically it doesn't break anything currently to have it, so not sure if it would be a priority.

**Comment by chapmanhk** [comment]
> what is a school-type triple?

@vishpillai123 it's just the three IDs we store on an institution: pdp_id, edvise_id, and legacy_id.

Together they answer “what kind of school is this?” The rule is: exactly one of the three should be set (PDP or Edvise Schema or Legacy—not more than one, not zero). It's basically just referring to “those three fields as a group,”

**Comment by vishpillai123** [comment]
> > what is a school-type triple?
> 
> @vishpillai123 it's just the three IDs we store on an institution: pdp_id, edvise_id, and legacy_id.
> 
> Together they answer “what kind of school is this?” The rule is: exactly one of the three should be set (PDP or Edvise Schema or Legacy—not more than one, not zero). It's basically just referring to “those three fields as a group,”

Ok got it makes sense! I feel like having two columns with "school_type", "school_id" makes a bit more intuitive sense to me, where school_type is limited to "pdp", "edvise", and "legacy". I'm assuming the ID is a random integer of some kind? 

Though I think this is fine to have the current design too.

**Comment by chapmanhk** [comment]
> > > what is a school-type triple?
> > 
> > 
> > @vishpillai123 it's just the three IDs we store on an institution: pdp_id, edvise_id, and legacy_id.
> > Together they answer “what kind of school is this?” The rule is: exactly one of the three should be set (PDP or Edvise Schema or Legacy—not more than one, not zero). It's basically just referring to “those three fields as a group,”
> 
> Ok got it makes sense! I feel like having two columns with "school_type", "school_id" makes a bit more intuitive sense to me, where school_type is limited to "pdp", "edvise", and "legacy". I'm assuming the ID is a random integer of some kind?
> 
> Though I think this is fine to have the current design too.

@vishpillai123 100% agree, and Bill wants this as well. We just already had is_pdp and pdp_id as columns, so it was following that format
